### PR TITLE
feat(diregapic): Add PHP client library DIREGAPIC support

### PR DESCRIPTION
**Pull request by miraleung**
Uses the same proto file as the [monolith PR](https://github.com/googleapis/gapic-generator/pull/3305/files), with minor import differences.

Notes:
- For the client diff, please  see the file `tests/Unit/ProtoTests/BasicDiregapic/out/src/Gapic/LibraryServiceGapicClient.php`.
- Generates the private `defaultTransport()` and `getSupportedTransports()` methods that the monolith produced, since they are overrides of gax-php's `GapicClientTrait`, and are called from that class.
- Please refer to #263 for the original, gRPC+REST goldens.
- Showcase and/or Bazel tests to come in a separate PR.
- The proto uses `BookResponse` and `ShelfResponse` instead of the more sensible `Book` and `Shelf` because of the way the unit test framework is set up. The corresponding protoc-generated PHP source doesn't actually exist - it's faked out by `tests/Unit/autoload.php`, which also doesn't handle enums.

### CODE REVIEW

File: src/Generation/GapicClientGenerator.php [hunk 1]
```diff
@@ -392,6 +402,45 @@ private function construct(): PhpClassMember
...
+            $transportConfigSampleValues['grpc'] = AST::arrayEllipsis();
+        }
+        // Set this value here, don't initialize it, so we can maintain alphabetical order
+        // for the resulting printed doc.
+        $transportConfigSampleValues['rest'] = AST::arrayEllipsis();
+        $transportConfigDocText =
+          PhpDoc::text(
+              'Configuration options that will be used to construct the transport. Options for',
+              'each supported transport type should be passed in a key for that transport. For example:',
+              PhpDoc::example(AST::block(
```
Comment (vam-google) [inline src/Generation/GapicClientGenerator.php:433]: Some strange formatting here too

File: src/Utils/Transport.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+declare(strict_types=1);
+
+namespace Google\Generator\Utils;
+
+class Transport
+{
+    // Consts are used instead of booleans, because there more transport options may be
+    // supported in the future (e.g. gRPC only).
+    public const GRPC_REST = 1;
+    public const REST_ONLY = 2;
```
Comment (vam-google) [inline src/Utils/Transport.php:26]: Why not just REST? Such that the constant names match the command like args and the cross-lang  DIREGAPIC spec in general?

Probably also add `GRPC`  value for completeness, even if it is not supported by the generator (but still can be used by the validation logic).

File: tests/Unit/ProtoTests/BasicDiregapic/out/src/Gapic/LibraryServiceGapicClient.php [hunk 3]
```diff
@@ -820,9 +819,8 @@ public function resumeOperation($operationName, $methodName = null)
      *           default this settings points to the default client config file, which is
      *           provided in the resources folder.
      *     @type string|TransportInterface $transport
-     *           The transport used for executing network requests. May be either the string
-     *           `rest` or `grpc`. Defaults to `grpc` if gRPC support is detected on the system.
-     *           *Advanced usage*: Additionally, it is possible to pass in an already
+     *           The transport used for executing network requests. At the moment, supports only
```
Comment (vam-google) [inline tests/Unit/ProtoTests/BasicDiregapic/out/src/Gapic/LibraryServiceGapicClient.php:822]: It looks like the old version of this file was supporting `rest` or `grpc`, making the generated client of GRPC_REST transport type. This is very confusing, but do I get it right that `grpc+rest` as generator parameter, results in creating PHP client, which can have either `grpc` or `rest` transport specified on initialization? I.e. the `grpc` and `rest` terms are heavily overloaded here and on the moment of generation and on the moment of instantiating generated client they mean slightly different things. 

I guess we do need to have a test for rest-only generated client, but did it overwrote the other important test case - `grpc+rest` (the original version of this `LibraryServiceGapicClient.php` file)?

**Review by vam-google — APPROVED** [review]
LGTM, but a few minor comments about resolving the `grpc` vs `grpc+rest` parameters confusion (in both places - as an argument to the generator and as an argument to the client on instantiation).

@bshaffer PTAL

File: rules_php_gapic/php_gapic.bzl [hunk 4]
```diff
@@ -61,6 +61,11 @@ def php_gapic_srcjar(
     if grpc_service_config:
         plugin_file_args[grpc_service_config] = "grpc_service_config"
 
+    if transport == None:
+        transport = "grpc"
```
[the remark this answers is not in the document]
Comment (miraleung) [reply rules_php_gapic/php_gapic.bzl:65]: Done.

File: src/Generation/GapicClientGenerator.php [hunk 5]
```diff
@@ -361,25 +362,34 @@ private function lroMethods(): Vector
...
+        'serviceName' => AST::access(AST::SELF, $this->serviceName()),
+        // TODO: Understand if this should be named 'serviceAddress' or 'apiEndpoint'?
+        'serviceAddress' => AST::concat(AST::access(AST::SELF, $this->serviceAddress()), ':', AST::access(AST::SELF, $this->servicePort())),
+        'clientConfig' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->clientConfigFilename}"),
+        'descriptorsConfigPath' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->descriptorConfigFilename}"),
+      ];
+
+        // TODO: Consolidate setting all the known array values together.
+        // We do this here to maintain the existing sensible ordering.
+        if ($this->serviceDetails->transportType === Transport::GRPC_REST) {
```
[the remark this answers is not in the document]
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:375]: Yes, updated `Transport.php` to error-out if `'grpc'` is given.

File: src/Generation/GapicClientGenerator.php [hunk 6]
```diff
@@ -361,25 +362,34 @@ private function lroMethods(): Vector
...
+            $clientDefaultValues['gcpApiConfigPath'] =
+          AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->grpcConfigFilename}");
+        }
+
+        $clientDefaultValues['credentialsConfig'] = AST::array([
+        'defaultScopes' => AST::access(AST::SELF, $this->serviceScopes()),
+      ]);
```
[the remark this answers is not in the document]
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:382]: Done.

File: src/Generation/GapicClientGenerator.php [hunk 7]
```diff
@@ -392,6 +402,45 @@ private function construct(): PhpClassMember
         $options = AST::var('options');
         $optionsParam = AST::param($this->ctx->type(Type::array()), $options, AST::array([]));
         $clientOptions = AST::var('clientOptions');
+
+        // Assumes there are only two transport types.
+        $isGrpcRest = $this->serviceDetails->transportType === Transport::GRPC_REST;
+
+        $restTransportDocText = 'At the moment, supports only `rest`.';
+        $grpcTransportDocText = 'May be either the string `rest` or `grpc`. Defaults to `grpc` if gRPC support is detected on the system.';
```
[the remark this answers is not in the document]
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:410]: This generated PhpDoc replicates the logic in the [monolith PR](https://github.com/googleapis/gapic-generator/pull/3305/files). @bshaffer could you please confirm if this is the appropriate doc text to use?

File: src/Generation/GapicClientGenerator.php (hunk 1 above)
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:433]: Done.

File: src/Utils/Transport.php (hunk 2 above)
Comment (miraleung) [reply src/Utils/Transport.php:26]: Done, leaving out `GRPC` since it's not used anywhere.

Technically, `REST` also isn't used anywhere and this could be a bool. I initially tried that out, but the ensuing logic (e.g. `isGrpcRest()`) implied that the negative state had only one value, which looked suboptimal from a readability and maintenance perspective.

File: tests/Unit/ProtoTests/BasicDiregapic/out/src/Gapic/LibraryServiceGapicClient.php (hunk 3 above)
Comment (miraleung) [reply tests/Unit/ProtoTests/BasicDiregapic/out/src/Gapic/LibraryServiceGapicClient.php:822]: The other goldens in `tests/Unit/ProtoTests/` will exercise the `grpc+rest` case. This documentation text replicates that in the [monolith PR](https://github.com/googleapis/gapic-generator/pull/3305/files), so happy to hear improvement suggestions.

File: src/Generation/GapicClientGenerator.php [hunk 8]
```diff
@@ -361,25 +362,34 @@ private function lroMethods(): Vector
 
     private function getClientDefaults(): PhpClassMember
     {
+        $clientDefaultValues = [
+        'serviceName' => AST::access(AST::SELF, $this->serviceName()),
+        // TODO: Understand if this should be named 'serviceAddress' or 'apiEndpoint'?
+        'serviceAddress' => AST::concat(AST::access(AST::SELF, $this->serviceAddress()), ':', AST::access(AST::SELF, $this->servicePort())),
+        'clientConfig' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->clientConfigFilename}"),
+        'descriptorsConfigPath' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->descriptorConfigFilename}"),
+      ];
```
Comment (bshaffer) [inline src/Generation/GapicClientGenerator.php:371]: nit: whitespace is off here. Should be:

```php
$clientDefaultValues = [
    'serviceName' => AST::access(AST::SELF, $this->serviceName()),
    // TODO: Understand if this should be named 'serviceAddress' or 'apiEndpoint'?
    'serviceAddress' => AST::concat(AST::access(AST::SELF, $this->serviceAddress()), ':', AST::access(AST::SELF, $this->servicePort())),
    'clientConfig' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->clientConfigFilename}"),
    'descriptorsConfigPath' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->descriptorConfigFilename}"),
];
```

File: src/Generation/GapicClientGenerator.php [hunk 9]
```diff
@@ -361,25 +362,34 @@ private function lroMethods(): Vector
 
     private function getClientDefaults(): PhpClassMember
     {
+        $clientDefaultValues = [
+        'serviceName' => AST::access(AST::SELF, $this->serviceName()),
+        // TODO: Understand if this should be named 'serviceAddress' or 'apiEndpoint'?
+        'serviceAddress' => AST::concat(AST::access(AST::SELF, $this->serviceAddress()), ':', AST::access(AST::SELF, $this->servicePort())),
```
Comment (bshaffer) [inline src/Generation/GapicClientGenerator.php:368]: nit: For readability can we split this up?

```php
$serviceAddress = AST::concat(
    AST::access(AST::SELF, $this->serviceAddress()),
    ':',
    AST::access(AST::SELF, $this->servicePort())
);
```

File: src/Generation/GapicClientGenerator.php [hunk 10]
```diff
@@ -361,25 +362,34 @@ private function lroMethods(): Vector
...
+        'clientConfig' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->clientConfigFilename}"),
+        'descriptorsConfigPath' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->descriptorConfigFilename}"),
+      ];
+
+        // TODO: Consolidate setting all the known array values together.
+        // We do this here to maintain the existing sensible ordering.
+        if ($this->serviceDetails->transportType === Transport::GRPC_REST) {
+            $clientDefaultValues['gcpApiConfigPath'] =
+          AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->grpcConfigFilename}");
```
Comment (bshaffer) [inline src/Generation/GapicClientGenerator.php:377]: nit: Whitespace is off here. Should be:

```php
$clientDefaultValues['gcpApiConfigPath'] =
    AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->grpcConfigFilename}");
```

File: src/Generation/GapicClientGenerator.php [hunk 11]
```diff
@@ -361,25 +362,34 @@ private function lroMethods(): Vector
...
+        }
+
+        $clientDefaultValues['credentialsConfig'] = AST::array([
+        'defaultScopes' => AST::access(AST::SELF, $this->serviceScopes()),
+      ]);
+        $clientDefaultValues['transportConfig'] = AST::array([
+        'rest' => AST::array([
+          'restClientConfigPath' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->restConfigFilename}"),
+        ])
+      ]);
```
Comment (bshaffer) [inline src/Generation/GapicClientGenerator.php:387]: Whitespace is off here too. I'll just make a general comment in my review to fix it everywhere, but 4 spaces and no tabs is our standard.

File: src/CodeGenerator.php [hunk 12]
```diff
@@ -45,6 +46,8 @@ class CodeGenerator
      *
      * @param string $descBytes The raw bytes of the proto descriptor, as generated using `protoc -o ...`
      * @param string $package The package name to generate.
+     * @param ?string $transport The type of transport to support, gRPC+REST by default (null).
```
Comment (bshaffer) [inline src/CodeGenerator.php:49]: for clarity

```suggestion
     * @param ?string $transport The type of transport to support, gRPC and REST by default (null).
```

File: src/Generation/GapicClientGenerator.php [hunk 13]
```diff
@@ -361,25 +362,34 @@ private function lroMethods(): Vector
 
     private function getClientDefaults(): PhpClassMember
     {
+        $clientDefaultValues = [
+        'serviceName' => AST::access(AST::SELF, $this->serviceName()),
+        // TODO: Understand if this should be named 'serviceAddress' or 'apiEndpoint'?
+        'serviceAddress' => AST::concat(AST::access(AST::SELF, $this->serviceAddress()), ':', AST::access(AST::SELF, $this->servicePort())),
+        'clientConfig' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->clientConfigFilename}"),
+        'descriptorsConfigPath' => AST::concat(AST::__DIR__, "/../resources/{$this->serviceDetails->descriptorConfigFilename}"),
```
Comment (bshaffer) [inline src/Generation/GapicClientGenerator.php:370]: For readability can we set these to local variables above?

```php
$clientConfigFilename = $this->serviceDetails->clientConfigFilename;
$descriptorConfigFilename = $this->serviceDetails->clientConfigFilename;

//...
    'clientConfig' => AST::concat(AST::__DIR__, "/../resources/{$clientConfigFilename}"),
    'descriptorsConfigPath' => AST::concat(AST::__DIR__, "/../resources/{$descriptorConfigFilename}"),
```

File: src/Generation/GapicClientGenerator.php [hunk 14]
```diff
@@ -392,6 +402,45 @@ private function construct(): PhpClassMember
...
+        $transportConfigSampleValues['rest'] = AST::arrayEllipsis();
+        $transportConfigDocText =
+          PhpDoc::text(
+              'Configuration options that will be used to construct the transport. Options for',
+              'each supported transport type should be passed in a key for that transport. For example:',
+              PhpDoc::example(AST::block(
+                  AST::assign(AST::var('transportConfig'), AST::array($transportConfigSampleValues))
+              ), null, true),
+              'See the',
+              AST::call($ctx->type(Type::fromName($isGrpcRest ? GrpcTransport::class : RestTransport::class), true), AST::method('build'))(),
```
Comment (bshaffer) [inline src/Generation/GapicClientGenerator.php:437]: nit: This is hard to follow... is it possible to split this out?

File: src/Generation/ServiceDetails.php [hunk 15]
```diff
@@ -114,16 +115,23 @@ class ServiceDetails
     /** @var bool *Readonly* Whether the service is deprecated. */
     public bool $isDeprecated = false;
 
+    /**
+     * @var int *Readonly* The transport type supported in this client. Valid options are in Transport.php.
```
Comment (bshaffer) [inline src/Generation/ServiceDetails.php:119]: if you want this to be readonly, I suggest using getters instead of public properties. 

I wouldn't mind a refactor of all these properties to use public getters without setters for this reason, as they're all labeled Readonly. Unless something else is meant by `Readonly`?

**Review by bshaffer — COMMENTED** [review]
Mostly nits on whitespace/readability with one comment on `readonly`. 

Overall the functionality and implementation looks great!

File: src/Generation/GapicClientGenerator.php (hunk 8 above)
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:371]: Done.

File: src/Generation/GapicClientGenerator.php (hunk 9 above)
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:368]: Done.

File: src/Generation/GapicClientGenerator.php (hunk 11 above)
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:387]: Done.

File: src/CodeGenerator.php (hunk 12 above)
Comment (miraleung) [reply src/CodeGenerator.php:49]: Done.

File: src/Generation/GapicClientGenerator.php (hunk 13 above)
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:370]: Done.

File: src/Generation/GapicClientGenerator.php (hunk 10 above)
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:377]: Done.

File: src/Generation/GapicClientGenerator.php (hunk 14 above)
Comment (miraleung) [reply src/Generation/GapicClientGenerator.php:437]: Done.

File: src/Generation/ServiceDetails.php (hunk 15 above)
Comment (miraleung) [reply src/Generation/ServiceDetails.php:119]: These patterns were in place before I took on this codebase; the `readonly` addition was intended to follow the existing structure. Happy to address these in general in separate PRs, but I'd like to keep this as-is for consistency here.

If you have general structural improvement suggestions, please let me know or file issues in this repo!  I'd be happy to tackle them over time. :)

File: src/Generation/GapicClientGenerator.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (bshaffer) [reply src/Generation/GapicClientGenerator.php:416]: "grpc+rest" is a value for the microgen, not for the GAPIC clients. GAPIC clients must be either grpc or rest at runtime (obviously, how can they make a call to the apis with both?). That is what is meant here.

**Comment by miraleung** [comment]
Added `defaultTransport` and `getSupportedTransports` as discussed offline.

**Comment by miraleung** [comment]
Friendly ping.

**Review by bshaffer — APPROVED** [review]
